protocol/linkpreview_unfurler_opengraph.go

```diff
@@ -38,6 +40,35 @@ func NewOpenGraphUnfurler(URL *neturl.URL, logger *zap.Logger, httpClient *http.
 	}
 }
 
+func GetFavicon(bodyBytes []byte) string {
+	htmlTokens := html.NewTokenizer(bytes.NewBuffer(bodyBytes))
+loop:
+	for {
+		tt := htmlTokens.Next()
+		switch tt {
+		case html.ErrorToken:
+			break loop
+		case html.StartTagToken:
+			t := htmlTokens.Token()
+			if t.Data != "link" {
+				continue
+			}
+			LinkContainsFavicon := false
+			for _, attribute := range t.Attr {
+				if attribute.Val == "icon" || attribute.Val == "shortcut icon" {
+					LinkContainsFavicon = true
+				}
+			}
+			for _, attribute := range t.Attr {
+				if LinkContainsFavicon && (strings.Contains(attribute.Val, ".ico") || strings.Contains(attribute.Val, ".png") || strings.Contains(attribute.Val, ".svg")) {
```

Comment:
I don't understand why we need to loop twice. Another problem is that the first loop could have found that there is in fact a favicon link, but it doesn't mean it's the same as the one currently under iteration in the second loop. It seems to me to be more precise to check here `attribute.Val` is either `icon` or `shortcut icon`, and then we can also eliminate the first loop.

Reply:
When we encounter a start tag, We loop the first loop to check all attributes of that start tag and make sure it's actually a favicon

The second loop we double check if the favicon resources is actually one of our supported types and then we return it.

Reply:
Notice the `case html.StartTagToken:`
The tag is the same in both loops.
But we don't know for a fact in what order the HTML attributes are added to the start tag, That's why I loop through all of them first to make sure it's in fact a link tag that's meant to be a favicon.

Reply:
I copied this implementation, Thanks Icaro!